tests/Incentives/test_coinbase.cpp

```diff
@@ -0,0 +1,157 @@
+ */
+
+#include <vector>
+
+#include "common/Constants.h"
+#include "common/MessageNames.h"
+#include "common/Serializable.h"
+#include "libCrypto/Schnorr.h"
+#include "libCrypto/Sha2.h"
+#include "libDirectoryService/DirectoryService.h"
+#include "libTestUtils/TestUtils.h"
+#include "libUtils/DataConversion.h"
+#include "libUtils/DetachedFunction.h"
+#include "libUtils/Logger.h"
+#include "libUtils/UpgradeManager.h"
+
+#define BOOST_TEST_MODULE coinbase
+#define BOOST_TEST_DYN_LINK
+#include <boost/test/unit_test.hpp>
+
+using namespace std;
+using namespace TestUtils;
+
+BOOST_AUTO_TEST_SUITE(test_coinbase)
+
+BOOST_AUTO_TEST_CASE(test_coinbase_correctness) {
+  INIT_STDOUT_LOGGER();
+  LOG_MARKER();
+
+  Mediator mediator(GenerateRandomKeyPair(), GenerateRandomPeer());
+  DirectoryService dummyDS(mediator);
+  Node dummyNode(mediator, 0, true);  // Unused in code anyways
+  Lookup dummyLookup(mediator, SyncType::NO_SYNC);
+  const uint64_t nonce{0};
+  auto dummyValidator = make_shared<Validator>(mediator);
+  AccountStore::GetInstance().Init();
+
+  AccountStore::GetInstance().AddAccount(Address(),
+                                         {TOTAL_COINBASE_REWARD, nonce});
+  AccountStore::GetInstance().UpdateStateTrieAll();
+
+  const uint32_t min_ds_size = 600;
+  const uint32_t min_num_shards = 5;
+  const uint32_t min_lookup_size = 5;
+
+  mediator.RegisterColleagues(&dummyDS, &dummyNode, &dummyLookup,
+                              dummyValidator.get());
+
+  auto dummy_shard_size = (DistUint8() % min_num_shards) + min_num_shards;
+
+  auto dummy_ds_size = DistUint8() % min_ds_size + min_ds_size;
+  auto dummy_shards = GenerateDequeueOfShard(dummy_shard_size);
+
+  auto dummy_ds_comm = GenerateRandomDSCommittee(dummy_ds_size);
+
+  auto dummy_lookup_num = DistUint8() % min_lookup_size + min_lookup_size;
+
+  LOG_GENERAL(INFO, "Shard size: " << dummy_shard_size
+                                   << " DS size: " << dummy_ds_size
+                                   << " Lookup Num: " << dummy_lookup_num);
+  VectorOfNode lookupNodes;
+
+  for (uint i = 0; i < dummy_lookup_num; i++) {
+    lookupNodes.emplace_back(GenerateRandomPubKey(), GenerateRandomPeer());
+  }
+
+  dummyLookup.SetLookupNodes(lookupNodes);
+
+  vector<bool> b1, b2;
+
+  *mediator.m_DSCommittee = dummy_ds_comm;
+  dummyDS.m_shards = dummy_shards;
+
+  const uint num_test_epoch =
+      1;  // Can be changed to NUM_FINAL_BLOCK_PER_POW for proper testing
+
+  for (uint i = 0; i < num_test_epoch; i++) {
+    uint j = 0;
+    for (const auto& shard : dummy_shards) {
+      b1 = GenerateRandomBooleanVector(shard.size());
+
+      b2 = GenerateRandomBooleanVector(shard.size());
+      dummyDS.SaveCoinbaseCore(b1, b2, shard, j++, i + 1);
+    }
+
+    b1 = GenerateRandomBooleanVector(dummy_ds_comm.size());
+    b2 = GenerateRandomBooleanVector(dummy_ds_comm.size());
+
+    dummyDS.SaveCoinbaseCore(b1, b2, dummy_ds_comm,
```

Comment:
Do you also need to add `CoinbaseReward::LOOKUP_REWARD` shard?

Reply:
InitCoinbase rewards them equally, so no  need